leaderboard/llvm_codesize/e_greedy/e_greedy.py

```diff
@@ -0,0 +1,115 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+"""ϵ-greedy policy for LLVM codesize."""
+import logging
+import os
+import random
+import sys
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from typing import NamedTuple
+
+from absl import flags
+
+from compiler_gym.envs import CompilerEnv, LlvmEnv
+
+# Import the ../eval_policy.py helper.
+sys.path.insert(0, os.path.dirname(os.path.realpath(__file__)) + "/..")
+from eval_policy import eval_policy  # noqa pylint: disable=wrong-import-position
+
+flags.DEFINE_float(
+    "epsilon", 0, "The ratio of patience to the size of the action space. "
+)
+FLAGS = flags.FLAGS
+
+
+class ActionReward(NamedTuple):
+    """An action -> reward tuple for a single step()."""
+
+    action: int
+    reward: float
+
+
+def select_best_action(env: CompilerEnv, executor: ThreadPoolExecutor) -> ActionReward:
+    """Determine the best action by trying all possible options and ranking them."""
+
+    def eval_action(fkd: CompilerEnv, action: int) -> ActionReward:
+        """Evaluate the given action."""
+        try:
+            _, reward, _, _ = fkd.step(action)
+        finally:
+            fkd.close()
+        return ActionReward(action=action, reward=reward)
+
+    # Create an ordered list of actions, best first, ordered by descding reward,
+    # using action index as a tie-breaker. Do this by creating n forks of the
+    # environment, one for every action, and evaluting the actions in parallel
+    # threads. Note that calls to fork() occur in the main thread for thread
+    # safety in case of environment restart.
+    futures = (
+        executor.submit(eval_action, env.fork(), action)
+        for action in range(env.action_space.n)
+    )
+    results = (future.result() for future in as_completed(futures))
+    results = sorted(
```

Comment:
Could have been faster if just scanning through the list and keep the best instead of sorting.

Reply:
Good point! Done :)